pkg/rbd/clone.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/pkg/util"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+)
+
+func (cs *ControllerServer) createCloneFromImage(ctx context.Context, req *csi.CreateVolumeRequest, cloneRbdVol *rbdVolume, cr *util.Credentials) error {
+
+	var err error
+
+	volumeSource := req.VolumeContentSource.GetVolume()
+	if volumeSource == nil {
+		return status.Error(codes.InvalidArgument, "volume content source cannot be empty")
+	}
+
+	volumeID := volumeSource.GetVolumeId()
+	if volumeID == "" {
+		return status.Error(codes.InvalidArgument, "volume content source volume ID cannot be empty")
```

Comment:
This should return `status.NotFound` in case cloning from a Volume is intended, but the reference to the SourceVolume is incorrect.

Reply:
will take care of it, same applies for the snapshot source also